Dataset of used cars from ebay

In [8]:
import pandas as pd
import numpy as np
autos=pd.read_csv("autos.csv",delimiter=',',encoding='Latin-1')

The dataset contains integers and objects. Some of the null values are assigned. 

In [9]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [10]:
def clean(c):
    c=c.replace("dateCrawled","date_crawled")
    c=c.replace("offerType","offer_type")
    c=c.replace("yearOfRegistration","registration_year")
    c=c.replace("vehicleType","Vehicle_type")
    c=c.replace("powerPS","power_ps")
    c=c.replace("monthOfRegistration","registration_month")
    c=c.replace("fuelType","fuel_type")
    c=c.replace("notRepairedDamage","unrepaired_damage")
    c=c.replace("dateCreated","ad_created")
    c=c.replace("nrOfPictures","number_of_pictures")
    c=c.replace("postalCode","postal_code")
    c=c.replace("lastSeen","last_seen")
    return c
new_columns=[]
for c in autos.columns:
    clean_c=clean(c)
    new_columns.append(clean_c)
autos.columns=new_columns
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest Vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  number_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00                   0   
1         bmw              nein  2016-04-04 00:00:00                   0   
2  volkswagen              nein  2016-03-26 00:00:00                   0   
3       smart              nein  2016-03-12 00:00:00                   0   
4        ford              nein  2016-04-01 00:00:00                   0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

The date set is visualised for uncleaned datas. The column labels are is camel case. These column labels are replaced by snake case objects.

In [11]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 19:38:20  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       Vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  number_of_pictures   postal_code  \
count                 50000             50000.0  50000.000000   
unique                   76                 NaN           NaN   
top     2016-04-03 00:00:00                 NaN           NaN   
freq                   1946                 NaN           NaN   
mean                    NaN                 0.0  50813.627300   
std                     NaN                 0.0  25779.747957   
min                     NaN                 0.0   1067.000000   
25%                     NaN                 0.0  30451.000000   
50%                     NaN                 0.0  49577.000000   
75%                     NaN                 0.0  71540.000000   
max                     NaN                 0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

The price and odometer column contains a non-numeric value. These column labels should be renamed and the values should be converted to int64 datatype inorder to manipulate the date.

In [12]:
autos.rename({"odometer":"odometer_km"},axis=1,inplace=True)
autos.loc[:,"price"]=autos.loc[:,"price"].str.replace("$","")
autos.loc[:,"odometer_km"]=autos.loc[:,"odometer_km"].str.replace("km","")
autos.loc[:,"price"]=autos.loc[:,"price"].str.replace(",","")
autos.loc[:,"odometer_km"]=autos.loc[:,"odometer_km"].str.replace(",","")
autos.loc[:,"price"]=autos.loc[:,"price"].astype(int)
autos.loc[:,"odometer_km"]=autos.loc[:,"odometer_km"].astype(int)
autos.loc[:,"odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

In [78]:
autos.loc[autos.loc[:,"price"]==0,'price']=np.nan

count    4.857900e+04
mean     1.012788e+04
std      4.880873e+05
min      1.000000e+00
25%      1.200000e+03
50%      3.000000e+03
75%      7.490000e+03
max      1.000000e+08
Name: price, dtype: float64

some of the data in the price column contains zero. sincce the price cannot be zero, it's been replaced as NaN(not a number).

In [79]:
autos.head(3)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offer_type   price   abtest Vehicle_type  registration_year    gearbox  \
0    Angebot  5000.0  control          bus               2004    manuell   
1    Angebot  8500.0  control    limousine               1997  automatik   
2    Angebot  8990.0     test    limousine               2009    manuell   

   power_ps   model  odometer_km  registration_month fuel_type       brand  \
0       158  andere       150000                   3       lpg     peugeot   
1       286     7er       150000                   6    benzin         bmw   
2       102    golf        70000                   7    benzin  volkswagen   

  unrepaired_damage           ad_created  number_of_pictures  postal_code  \
0              nein  2016-03-26 00:00:00                   0        79588   
1              nein  2016-04-04 00:00:00                   0        71034   
2              nein  2016-03-26 00:00:00                   0        35394   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

In [80]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price                 48579 non-null float64
abtest                50000 non-null object
Vehicle_type          44905 non-null object
registration_year     50000 non-null int64
gearbox               47320 non-null object
power_ps              50000 non-null int64
model                 47242 non-null object
odometer_km           50000 non-null int64
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
number_of_pictures    50000 non-null int64
postal_code           50000 non-null int64
last_seen             50000 non-null object
dtypes: float64(

In [110]:
autos.loc[:,"registration_year"].describe()

count    49976.000000
mean      2003.366836
std          7.693943
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2019.000000
Name: registration_year, dtype: float64

some of the outliners are as follows:
The minimum value is set to 1000.
The maximum value is set to 9999.

The Car must be registered between 1900 and 2020.

In [111]:
autos_selected=autos[(autos.loc[:,"registration_year"]>1900)&(autos.loc[:,"registration_year"]<2020)]
autos_selected.shape

(49976, 20)

24 values are outlined in the column- registration_year

In [113]:
autos_selected.loc[:,"registration_year"].value_counts(normalize=True)
autos=autos_selected

In [117]:
autos.loc[:,"brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [135]:
count=autos.loc[:,"brand"].value_counts()
percent=(count/autos.loc[:,"brand"].unique().shape)*100
print(percent)

volkswagen        26705.0
opel              13645.0
bmw               13570.0
mercedes_benz     11830.0
audi              10707.5
ford               8697.5
renault            6010.0
peugeot            3640.0
fiat               3270.0
seat               2350.0
skoda              1962.5
mazda              1892.5
nissan             1885.0
smart              1752.5
citroen            1750.0
toyota             1542.5
sonstige_autos     1355.0
hyundai            1220.0
volvo              1142.5
mini               1060.0
mitsubishi         1010.0
honda               997.5
kia                 890.0
alfa_romeo          822.5
porsche             735.0
suzuki              732.5
chevrolet           707.5
chrysler            452.5
dacia               322.5
daihatsu            320.0
jeep                272.5
subaru              270.0
land_rover          247.5
daewoo              197.5
saab                197.5
jaguar              192.5
trabant             192.5
rover               172.5
lancia      

since all the brands are greater than 5% of the total count, we can take all the brands for aggregation

In [191]:
mean_dict={}
for i in autos.loc[:,"brand"].unique():
    brand_bool=autos.loc[:,"brand"]==i
    mean=autos.loc[brand_bool,"price"].mean()
    mean_dict[i]=mean
bmp_series=pd.Series(mean_dict)
df=pd.DataFrame(bmp_series,columns=["mean_price"])
print(df)

                  mean_price
alfa_romeo       4041.844237
audi             9212.930662
bmw              8493.857251
chevrolet        6620.069091
chrysler         3519.384615
citroen         43578.909621
dacia            5897.736434
daewoo           1079.342105
daihatsu         1628.426230
fiat            12566.541568
ford             7305.141844
honda            4000.139175
hyundai          5371.792961
jaguar          11525.554054
jeep            11590.214953
kia              5889.298551
lada             2647.724138
lancia           3182.000000
land_rover      18934.272727
mazda            4059.059540
mercedes_benz   30025.542249
mini            10541.566986
mitsubishi       3386.911839
nissan           4669.385965
opel             5281.321706
peugeot          3065.611888
porsche         45624.271777
renault          2431.195699
rover            1586.492308
saab             3183.493671
seat             4320.168662
skoda            6353.544872
smart            3518.102305
sonstige_autos

The mean price of each car brand is calculated using aggregation. using the mean price and percentage of cars registered, we can analyse the cars selled despite their price.

In [201]:
df_sorted=df.sort_values("mean_price",ascending=False)
q=df_sorted.head().index
print(q)

Index(['porsche', 'sonstige_autos', 'citroen', 'volvo', 'mercedes_benz'], dtype='object')


In [202]:
'''mean_mileage={}
for i in q:
    mileage_bool=autos.loc[:,"brand"]==i
    mean_mil=autos.loc[mileage_bool,""].mean()
    mean_mileage[i]=mean_mil
mileage_series=pd.Series(mean_mileage)
new_df=pd.DataFrame([bmp_series,mileage_series],columns=["mean_price","mean_mileage"])
print(new_df)'''

'mean_mileage={}\nfor i in q:\n    mileage_bool=autos.loc[:,"brand"]==i\n    mean_mil=autos.loc[mileage_bool,""].mean()\n    mean_mileage[i]=mean_mil\nmileage_series=pd.Series(mean_mileage)\nnew_df=pd.DataFrame([bmp_series,mileage_series],columns=["mean_price","mean_mileage"])\nprint(new_df)'